src/Bicep.Core.Samples/Files/InvalidResources_CRLF/main.diagnostics.bicep

```diff
@@ -102,7 +102,6 @@ resource foo 'Microsoft.Foo/foos@2020-02-02-alpha'= if {
 //@[13:50) [BCP081 (Warning)] Resource type "Microsoft.Foo/foos@2020-02-02-alpha" does not have types available. (CodeDescription: none) |'Microsoft.Foo/foos@2020-02-02-alpha'|
 //@[55:56) [BCP018 (Error)] Expected the "(" character at this location. (CodeDescription: none) |{|
   name: 'foo'
-//@[8:13) [BCP121 (Error)] Resources: "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo", "foo" are defined with this same name in a file. Rename them or split into different modules. (CodeDescription: none) |'foo'|

```

Comment:
>BCP121 [](http://example.com/codeflow?start=11&length=6)

Why did this error go away?

Reply:
I'd bet that `ResourceDefinition.EqualityComparer` is not compatible with the ResourceMetadata as scope.

Reply:
to be more precise - this line: https://github.com/Azure/bicep/blob/ab97e2d051a69397729f09cda71e309cb0f6a0e3/src/Bicep.Core/Utils/ResourceDefinition.cs#L41

Reply:
I think the difference is that prior to my change, we enumerated `model.Root.GetAllResourceDeclarations()` for the list of resources to analyze, which uses `ResourceSymbolVisitor.GetAllResources()`. This visitor also returns **unbound** resource symbols, but the `ResourceMetadataCache` class I added uses the binder to enumerate resource symbols (meaning it only contains **bound** resource symbols). The resource examples in this file have both duplicate **names** and duplicate **symbolic names** - because of the duplicate **symbolic names**, only the first `foo` resource declaration gets bound; the duplicated ones are not enumerated.

It's going to a bit tricky to try and fix this - either I'd have to add unbound symbols to the `ResourceMetadataCache`, or I'd have to stop using `ResourceMetadata` as a key on the scope data dictionary. I'm not keen on the former as I'm wanting to treat `ResourceMetadata` sort of like an IR for template generation.

Reply:
I'm confused about something. The code generation today is done from declared symbols and not bound symbols. If you want to use this as an IR, wouldn't you have to solve this anyway? Also, why can't you have different methods for requesting data for different contexts (some would return declared or just unbound symbols and others could return the bound ones).

Reply:
> I'm confused about something. The code generation today is done from declared symbols and not bound symbols. If you want to use this as an IR, wouldn't you have to solve this anyway? Also, why can't you have different methods for requesting data for different contexts (some would return declared or just unbound symbols and others could return the bound ones).

Declared but unbound symbols means something went wrong during name binding (e.g. duplicate symbols) - so code generation is blocked. So essentially code generation only uses bound declared symbols.

Reply:
That turned out to be a very easy change to make!